tooling/templatize/pkg/pipeline/types.go

```diff
@@ -14,25 +16,170 @@ type Pipeline struct {
 }
 
 type ResourceGroup struct {
-	Name         string  `yaml:"name"`
-	Subscription string  `yaml:"subscription"`
-	AKSCluster   string  `yaml:"aksCluster,omitempty"`
-	Steps        []*Step `yaml:"steps"`
+	Name         string `yaml:"name"`
+	Subscription string `yaml:"subscription"`
+	AKSCluster   string `yaml:"aksCluster,omitempty"`
+	Steps        []Step `yaml:"steps"`
+}
+
+func (rg *ResourceGroup) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	rawRg := &struct {
+		Name         string    `yaml:"name"`
+		Subscription string    `yaml:"subscription"`
+		AKSCluster   string    `yaml:"aksCluster,omitempty"`
+		Steps        []rawStep `yaml:"steps"`
+	}{}
+	if err := unmarshal(&rawRg); err != nil {
+		return err
+	}
+	rg.Name = rawRg.Name
+	rg.Subscription = rawRg.Subscription
+	rg.AKSCluster = rawRg.AKSCluster
+	rg.Steps = make([]Step, len(rawRg.Steps))
+	for i, rawStep := range rawRg.Steps {
+		switch rawStep.meta.Action {
+		case "Shell":
+			rg.Steps[i] = &ShellStep{}
+		case "ARM":
+			rg.Steps[i] = &ARMStep{}
+		default:
+			return fmt.Errorf("unknown action type %s", rawStep.meta.Action)
+		}
+		err := rawStep.unmarshal(rg.Steps[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
 type outPutHandler func(string)
 
-type Step struct {
-	Name            string     `yaml:"name"`
-	Action          string     `yaml:"action"`
+type StepMeta struct {
+	Name      string   `yaml:"name"`
+	Action    string   `yaml:"action"`
+	DependsOn []string `yaml:"dependsOn,omitempty"`
+}
+
+func (m *StepMeta) StepName() string {
+	return m.Name
+}
+
+func (m *StepMeta) ActionType() string {
+	return m.Action
+}
+
+func (m *StepMeta) Dependencies() []string {
+	return m.DependsOn
+}
+
+type rawStep struct {
+	meta      *StepMeta
+	unmarshal func(interface{}) error
+}
+
+func (msg *rawStep) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	msg.meta = &StepMeta{}
+	if err := unmarshal(msg.meta); err != nil {
+		return err
+	}
+	msg.unmarshal = unmarshal
+	return nil
+}
+
+func (msg *rawStep) Unmarshal(v interface{}) error {
+	return msg.unmarshal(v)
+}
+
+type Step interface {
+	StepName() string
+	ActionType() string
+	Description() string
+	Dependencies() []string
+}
+
+func NewShellStep(name string, command string) *ShellStep {
+	return &ShellStep{
+		StepMeta: StepMeta{
+			Name:   name,
+			Action: "Shell",
+		},
+		Command: command,
+	}
+}
+
+type ShellStep struct {
+	StepMeta   `yaml:",inline"`
+	Command    string     `yaml:"command,omitempty"`
+	Variables  []Variable `yaml:"variables,omitempty"`
+	DryRun     DryRun     `yaml:"dryRun,omitempty"`
+	outputFunc outPutHandler
```

Comment:
This smells a tiny bit - and makes static validation of your code harder by storing function pointers with data, which renders things like `cmp.Diff()` useless. Who creates the output handler, is it a runtime concern? Why does it need to be attached to the step as opposed to somehow passed in to a `Run()`?

Reply:
i agree. it has been here already so no change. let's tackle this in a followup PR. it should be part of the Run func somehow